src/tui/mod.rs

```diff
@@ -0,0 +1,144 @@
+mod terminal;
+
+use crate::async_buffer_redirect::AsyncBufferRedirect;
+use crate::ShutdownHandle;
+use ansi_to_tui::IntoText;
+use crossterm::event::Event;
+use crossterm::event::EventStream;
+use crossterm::event::KeyCode;
+use crossterm::event::KeyModifiers;
+use crossterm::event::MouseEventKind;
+use miette::miette;
+use miette::IntoDiagnostic;
+use miette::WrapErr;
+use ratatui::prelude::Buffer;
+use ratatui::prelude::Rect;
+use ratatui::widgets::Paragraph;
+use ratatui::widgets::Widget;
+use ratatui::widgets::Wrap;
+use std::cmp::min;
+use tokio::io::AsyncReadExt;
+use tokio::io::DuplexStream;
+use tokio_stream::StreamExt;
+
+#[derive(Default)]
+struct Tui {
+    quit: bool,
+    scrollback: Vec<u8>,
+    // TODO(evan): Follow output when scrolled to bottom
+    scroll_offset: usize,
+}
+
+/// TODO(evan): Document
+pub async fn run_tui(
+    mut shutdown: ShutdownHandle,
+    mut ghci_reader: DuplexStream,
+) -> miette::Result<()> {
+    let mut tracing_reader = AsyncBufferRedirect::stderr()
+        .into_diagnostic()
+        .wrap_err("Failed to capture stderr")?;
+
+    let mut ghci_buffer = [0; 1024];
+    let mut tracing_buffer = [0; 1024];
+
+    let mut terminal = terminal::enter()?;
+
+    let mut tui = Tui::default();
+
+    let mut event_stream = EventStream::new();
+
+    while !tui.quit {
+        let mut render_result = Ok(());
+        terminal
+            .draw(|frame| {
+                let area = frame.size();
+                let buffer = frame.buffer_mut();
+                render_result = render(&tui, area, buffer);
+            })
+            .into_diagnostic()
+            .wrap_err("Failed to draw to terminal")?;
+        render_result?;
+
+        tokio::select! {
+            _ = shutdown.on_shutdown_requested() => {
+                tui.quit = true;
+            }
+
+            output = ghci_reader.read(&mut ghci_buffer) => {
+                let n = output
+                    .into_diagnostic()
+                    .wrap_err("Failed to read bytes from GHCi into TUI buffer")?;
+                if n == 0 {
+                    tui.quit = true;
+                } else {
+                    tui.scrollback.extend(ghci_buffer);
+                    ghci_buffer = [0; 1024];
+                }
+            }
+
+            output = tracing_reader.read(&mut tracing_buffer) => {
+                output
+                    .into_diagnostic()
+                    .wrap_err("Failed to read bytes from tracing into TUI buffer")?;
+                tui.scrollback.extend(tracing_buffer);
+                tracing_buffer = [0; 1024];
+            }
```

Comment:
Filling and zeroing separate buffers because I can't hand out multiple exclusive references to `tui.scrollback`. There's probably a better (more efficient and/or idiomatic) way to do this?